### PR TITLE
Add metrics collection

### PR DESCRIPTION
**Pull request by tarampampam**
| Q             | A
| ------------- | ---
| Bug fix?      | No
| New feature?  | Yes

### Added

- Interface `MetricsCollectionInterface`
- Formatters `JsonFormatter` and `PrometheusFormatter` now supports `MetricsCollectionInterface`
- Metric groups can be defined in `metrics.metric_classes`
- `MetricsManagerInterface::iterate(array $abstracts)` method

### Changed

- `MetricsManager` can works with metric groups now
- `MetricsManager::addFactory` allows to register fabric for object that implements `MetricsGroupInterface`
- `MetricsManagerInterface::make` now return `MetricInterface` or `MetricsGroupInterface` (method signature changed)

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I wrote unit tests for my code
- [x] I have made changes in [CHANGELOG.md](https://github.com/avto-dev/app-metrics-laravel/blob/master/CHANGELOG.md) file

### CODE REVIEW

File: src/Formatters/JsonFormatter.php [hunk 1]
```diff
@@ -33,27 +34,43 @@ public function format(iterable $metrics, int $options = \JSON_UNESCAPED_UNICODE
...
-                $item = [
-                    'value' => $metric->value(),
-                ];
-
-                if ($metric instanceof HasDescriptionInterface) {
-                    $item['description'] = $metric->description();
+            if ($metric instanceof MetricsGroupInterface) {
+                foreach ($metric->metrics() as $collection_item) {
+                    if ($collection_item instanceof MetricInterface) {
+                        $result[$collection_item->name()] = (object) $this->metricToArray($collection_item);
```
Comment (Reallife) [inline src/Formatters/JsonFormatter.php]: Collisions if metrics in group have same names

**Review by Reallife — CHANGES_REQUESTED** [review]

File: src/Formatters/JsonFormatter.php (hunk 1 above)
Comment (Reallife) [reply src/Formatters/JsonFormatter.php]: fixed

**Review by Reallife — APPROVED** [review]

**Comment by tarampampam** [comment]
@Reallife cbe36c72ce0916f403bff82f7c38a4db92012afc - wtf is this? I see backward incompatible changes and json formatter now returns array instead object with `property = metric name`. Collision can be only with invalid package configuration. Why you do this and publish release without my review?

**Comment by Reallife** [comment]
@tarampampam Look, for example you have group of metrics:
```php
class SomeMetricGroup implements MetricsGroupInterface
{
    public function metrics(): iterable
    {
        foreach (['foo','bar'] as $label_value) {
            yield new SomeMetric($label_value);
        }
    }
}
```
and you have a metric:
```php
class SomeMetric implements MetricInterface, HasLabelsInterface
{
    protected $label_value;

    public function __construct(string $label_value)
    {
        $this->label_value    = $label_value;
    }

    public function labels(): array
    {
        return [
            'label_key' => $label_value,
        ];
    }
    
    public function name(): string
    {
        return 'some_metric';
    }

    public function value(): int
    {
        return 1;
    }
}
```
Metrics classes array in package config looks like (it's wrong config?):
```php
return [
    ...
    'metric_classes' => [
        App\Metrics\SomeMetricGroup::class,
    ],
   ... 
];
```

With old version of JsonFormatter you'll get something like that:
```json
{
  "some_metric": {
    "value": 1,
    "labels": {
      "label_key": "bar"
    }
  }
}
```

Metric with label's value `foo` was replaced with metric with label's value `bar`.

**Pull request merged**